src/auto-evo/simulation/food_source/HeterotrophicFoodSource.cs

```diff
@@ -1,21 +1,17 @@
-﻿using System.Linq;
-
-public class HeterotrophicFoodSource : FoodSource
+﻿public class HeterotrophicFoodSource : FoodSource
 {
     private readonly Compound oxytoxy = SimulationParameters.Instance.GetCompound("oxytoxy");
 
     private MicrobeSpecies prey;
     private Patch patch;
     private float preySpeed;
-    private float preySize;
     private float totalEnergy;
 
     public HeterotrophicFoodSource(Patch patch, MicrobeSpecies prey)
     {
         this.prey = prey;
         this.patch = patch;
         preySpeed = prey.BaseSpeed;
-        preySize = prey.Organelles.Organelles.Sum(organelle => organelle.Definition.HexCount);
```

Comment:
It would help to compute these values just once instead of each time where the property is referenced it gets recomputed.

Reply:
Does the property here not work that way? https://github.com/Revolutionary-Games/Thrive/pull/2578/files/c3a05d9d6569411931c6eeafb52416763817d92e#diff-ef5f11f4e06e07f19c774dfe97a969568398f94e92a1fcff255cde04ac6e1aa7R40

If not, then I might need some C# tips so that it does.

Reply:
It does not. Literally each time when that property appears, the entire code for calculating it is re-ran, which as it looks pretty expensive of an operation is not very good. There's no inbuilt caching or anything like that in C# properties.

Properties can have a backing field and a dirty flag so that they can be recomputed only when necessary but here as it depends on the organelle list it's going to be a bit perhaps tricky to setup the callbacks to set the dirty flag properly after those change.

Reply:
I've gone ahead and reused the calculation for the prey size. I think the danger is too high to put it on the property altogether given how often the organelles are modified in the code so far, and while technically this is called O(n^2) times (each species attempting to eat all others), I'd be very concerned if N exceeded 10-20 in any given patch.

Reply:
> Properties can have a backing field and a dirty flag so that they can be recomputed only when necessary

Personally I love using the following syntax in case where a dirty flag is required for a caching field:

```
private type? value;
public type Property => value ??= calculation();

public void DirtyProperty() => value = null;
```

This can combine the dirty flag and the value itself into one concise variable. Of course this is only applicable when null is not a valid value.

Reply:
Having taken a while to look at this, I think the best next step is to actually pull this variable back off of the MicrobeSpecies class, and add it to a new wrapper used solely by the auto-evo. This information is auto-evo's domain, and a wrapper pattern would allow us to reuse a lot more calculations, including very esoteric ones like "number of times toxin hits you before you die"

Reply:
Except the same code is used to check if something can engulf something.
It would be highly silly to decouple the auto-evo from the actual game mechanics related calculation...

Reply:
The hex size calculation is separate from base hex size. Base hex size exists only on this branch.

Reply:
What? Why is it different? The question for "can I engulf this?" should be 100% the same between auto-evo and gameplay code. Why it isn't?

Reply:
Because a Microbe can have a variable amount of organelles (it could be anywhere in the process of division), whereas a MicrobeSpecies cannot. Yes, they both divide the result by 2 if the species is bacteria, but other than that they count different lists.

Reply:
Ah yes, of course, I forgot about that.
Anyway I think this is really a *property* of the species and belongs there.

Reply:
I'll make my attempt to convince you of open closed principle at a later time. For now, I have no intent on moving this code anywhere.

Reply:
I think someone, at some point is much more likely to reimplement the hex size computation as they can't find it when it is put in the wrong place, which is just in the depths of the auto-evo system.
Anyway, I won already as you put the code in the MicrobeSpecies class already.